audio/asap/man/sap2ntsc.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "SAP2NTSC" 1 "2024-03-19" "5.3.0" "SlackBuilds.org"
.SH NAME
sap2ntsc \- convert SAP Atari 8-bit chiptune files to NTSC timing
.\" RST source for sap2ntsc(1) man page. Convert with:
.
.\" rst2man.py sap2ntsc.rst > sap2ntsc.1
.
.SH SYNOPSIS
.sp
\fBsap2ntsc\fP \fBinputfile\fP [\fI\&...\fP]
.SH DESCRIPTION
.sp
\fBsap2ntsc\fP converts an Atari 8\-bit chiptune in SAP format from
PAL timing to NTSC. Not all SAP files can be converted: ones that
appear to already be NTSC, or ones that use the FASTPLAY option, are
unconvertible.
.sp
\fBBeware\fP: each \fBinputfile\fP is overwritten without confirmation.
If you need backups of the original files, you should make copies
\fIbefore\fP running \fBsap2ntsc\fP\&.
.sp
Non\-SAP chiptune files (e.g. RMT, CMC) are not supported.
.SH OPTIONS
.INDENT 0.0
.TP
.B \fB\-h\fP, \fB\-\-help\fP
Show built\-in help.
.TP
.B \fB\-v\fP, \fB\-\-version\fP
Show version number.
.UNINDENT
.SH COPYRIGHT
.sp
See the file /usr/doc/asap\-5.3.0/COPYING for license information.
.SH AUTHORS
.sp
The ASAP suite was written by Piotr Fusik, with contributions from many
others (see the website for details).
.sp
This man page written for the SlackBuilds.org project
by B. Watson, and is licensed under the WTFPL.
.SH SEE ALSO
.sp
\fBasapconv\fP(1), \fBasap\-sdl\fP(1), \fBasap\-mplayer\fP(1), \fBchksap.pl\fP(1), \fBsap2txt\fP(1)
.sp
The ASAP website: \fI\%https://asap.sourceforge.net/\fP
.\" Generated by docutils manpage writer.
.


ChangeLog.txt
Sat Mar 23 01:18:40 UTC 2024
academic/cryptominisat: Updated for version 5.11.21.
academic/ent: Use https URLs.
audio/asap: Updated for version 6.0.3.
audio/aseqview: Fix download URL.
audio/clementine: Fix build with protobuf3.
audio/connie: Fix homepage.
audio/jack1: Updated for version 0.126.0.
audio/podget: Updated for version 0.9.3.
audio/puddletag: Version bump to 2.3.0
audio/rubberband: Updated for version 3.3.0.
audio/tap_plugins: Fix download URL.
audio/vocoder-ladspa: Fix URLs.
audio/vocoder: Fix URLs.
audio/volumeicon: Fix download URL.
audio/xmms-jack: Fix URLs.
desktop/cwm: Updated for version 7.4.
desktop/icewm: Updated for version 3.4.6.
desktop/nwg-displays: Updated for version 0.3.15.
desktop/nwg-displays: Updated for version 0.3.16.
desktop/nwg-drawer: Updated for version 0.4.7.
desktop/nwg-hello: Updated for version 0.1.8.
desktop/nwg-look: Updated for version 0.2.7.
desktop/nwg-menu: Updated for version 0.1.3.
desktop/nwg-panel: Updated for version 0.9.26.
desktop/salmon: Fix URLs.
desktop/wmcpuwatch: Fix download URL, add doinst.sh.
desktop/wmnet: Fix URLs.
desktop/wofi-pass: Updated for version 24.0.2
desktop/xsecurelock: Make pandoc-bin an optional dependency
development/SQLAlchemy: Updated for version 2.0.28.
development/apache-log4cxx: Fix download URL.
development/as31: Fix URLs.
development/aws-cdk: Updated for version 2.133.0.
development/bbcsdl: Updated script
development/bsdiff: Fix download URL.
development/devhelp: compress man page.
development/github-desktop: Updated for version 3.3.10.
development/hexe: Fix homepage.
development/jdk: Updated for version 8u391.
development/jupyter_packaging: Update for 0.12.3
development/mongodb-compass: Updated for version 1.42.3.
development/msbasic2ascii: Fix URLs.
development/nvidia-container-toolkit: Added (NVidia container toolkit).
development/pgmodeler: Updated for version 1.1.1.
development/postman: Updated for version 10.24.3
development/protobuf3: Updated for version 26.0.
development/pycharm: Updated for version 2023.3.5.233.15026.15.
development/remake: Updated for version 4.3+dbg_1.6_4
development/uftrace: Updated for version 20240313_1919a15
development/vscode-bin: Updated for version 1.87.2.
development/wxHexEditor: Update script.
games/advancemame: Updated for version 3.10.
games/alephone: Updated for version 20240119.
games/citra: Updated for version 0.2090.
games/joy2key: Fix download URL.
games/marathon-data: Updated for version 20240119.
games/marathon-infinity-data: Updated for version 20240119.
games/marathon2-data: Updated for version 20240119.
games/pioneer: Updated for version 20240314.
gis/OWSLib: Updated for version 0.30.0.
graphics/MaterialX: Fix github URLs.
graphics/brlcad: updated for version 7.38.2
graphics/converseen: Updated for version 0.12.2.0.
graphics/converseen: Updated for version 0.12.2.1.
graphics/gifsicle: Updated for version 1.95.
graphics/imv: Updated for version 4.5.0, add libjxl support.
graphics/kuickshow: Updated for version 20240316_791359a.
graphics/mozjpeg: Updated for version 4.1.5.
graphics/pqiv: Updated for version 2.13.1.
graphics/tgif-QPL: Fix download URL.
graphics/vuescan: Updated for version 9.8.30.
libraries/coeurl: Add patch.
libraries/girara: Updated for version 0.4.2.
libraries/libaec: Updated for version 1.1.3.
libraries/libminizip: Update for 1.2.13 (+new maintainer)
libraries/libwpe: Updated for version 1.16.0.
libraries/openimagedenoise: Added (High-Performance Denoising Library).
libraries/openvkl: Added (Intel Open Volume Kernel Library).
libraries/ospray: Added (Ray Tracing Engine).
libraries/ospray_studio: Added (interactive visualization and ray tracing application).
libraries/protobuf-c: Fix build with new protobuf.
libraries/rkcommon: Added (Intel RenderKit Utilities).
libraries/simde: Added (Implementations of SIMD).
libraries/tree-sitter: Updated for version 0.22.2
libraries/unixODBC: Updated for version 2.3.12.
libraries/webkit2gtk4.1: Fix BUILD.
libraries/webkit2gtk4.1: Fix build on x86.
libraries/webkit2gtk4.1: Updated for version 2.44.0.
libraries/webkit2gtk: Fix build on x86.
libraries/webkit2gtk: Updated for version 2.44.0.
misc/bitwarden-desktop: revert to version 2024.1.0
misc/ghostpcl: Updated for version 10.03.0.
misc/ramdefrag: Fix homepage URL & maintainer email
misc/xdelta: Fix download URL.
multimedia/guvcview: Updated for version 2.1.0.
multimedia/lightspark: Updated for version 0.8.7.
multimedia/podcastparser: move to libraries.
multimedia/sickchill: Support newer SQLAlchemy
multimedia/vkeybd: Fix download URL.
network/csync: Fix download URL.
network/cyrus-imapd: Updated for version 3.2.12.
network/discord: Version bump to 0.0.46
network/dropbox: Updated for version 195.4.4995.
network/librewolf: Updated for version 123.0.1
network/midori: Removed (abandoned upstream).
network/mumble-server: Added ICE option
network/mumble-server: update build number
network/mumble: fix gtav plugin 32-bit compile
network/nheko: Update script.
network/nyxt: Updated for version 3.11.5.
network/opera: Updated for version 108.0.5067.29.
network/opera: Updated for version 108.0.5067.40.
network/signal-desktop: Updated for version 7.3.0.
network/tcpbench: Added (OpenBSD tcpbench).
network/telegram: Updated for version 4.15.2.
network/thorium-browser: Updated for version 122.0.6261.132.
network/tor-browser: Updated for version 13.0.13.
network/unbound: Updated for version 1.19.3.
network/vimb: Updated for version 3.7.0.
network/vivaldi: Updated for version 6.6.3271.53.
network/xmrig: Updated for version 6.21.1.
network/zabbix_agent2: Added (new generation of Zabbix agent).
network/zoom-linux: Updated for version 5.17.11.3835
office/mupdf: Updated for version 1.23.11.
office/zathura: Updated for version 0.5.4.
perl/perl-Archive-Zip: Updated for version 1.68.
perl/perl-Clone: Updated for version 0.46.
perl/perl-HTTP-CookieJar: Added (Perl Module).
perl/perl-Net-LibIDN2: Added (Perl bindings for GNU Libidn2).

audio/asap/sap2ntsc.c
/*
 * sap2ntsc.c - convert PAL SAP files to NTSC
 *
 * Copyright (C) 2012-2019  Piotr Fusik
 *
 * This file is part of ASAP (Another Slight Atari Player),
 * see http://asap.sourceforge.net
 *
 * ASAP is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * ASAP is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ASAP; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <stdarg.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* public API doesn't allow us to set NTSC or FASTPLAY */
#include "asap.c"

static void fatal_error(const char *format, ...)
{
	va_list args;
	va_start(args, format);
	fprintf(stderr, "sap2ntsc: ");
	vfprintf(stderr, format, args);
	fputc('\n', stderr);
	va_end(args);
	exit(1);
}

static void process_file(const char *filename)
{
	/* check filename */
	const char *ext = strrchr(filename, '.');
	if (ext == NULL || strcasecmp(ext, ".sap") != 0)
		fatal_error("%s: filename must be *.sap", filename);

	/* read file */
	FILE *fp = fopen(filename, "rb");
	if (fp == NULL)
		fatal_error("cannot open %s", filename);
	static unsigned char module[ASAPInfo_MAX_MODULE_LENGTH];
	int module_len = fread(module, 1, sizeof(module), fp);
	fclose(fp);

	/* parse file */
	ASAPInfo *info = ASAPInfo_New();
	if (info == NULL)
		fatal_error("out of memory");
	if (!ASAPInfo_Load(info, filename, module, module_len))
		fatal_error("%s: unsupported file", filename);

	/* check if conversion possible */
	if (ASAPInfo_IsNtsc(info))
		fatal_error("%s: is already NTSC", filename);
	if (ASAPInfo_GetPlayerRateScanlines(info) != 312)
		fatal_error("%s: uses FASTPLAY", filename);

	/* do the conversion */
	ASAPWriter *writer = ASAPWriter_New();
	if (writer == NULL)
		fatal_error("out of memory");
	info->ntsc = true;
	info->fastplay = 262;
	for (int i = 0; i < ASAPInfo_GetSongs(info); i++) {
		int duration = ASAPInfo_GetDuration(info, i);
		if (duration > 0)
			ASAPInfo_SetDuration(info, i, (int) (duration * (1773447 / 1789772.5 * 262 / 312)));
	}
	static unsigned char output[ASAPInfo_MAX_MODULE_LENGTH];
	ASAPWriter_SetOutput(writer, output, 0, sizeof(output));
	int output_len = ASAPWriter_Write(writer, filename, info, module, module_len, true);
	ASAPWriter_Delete(writer);
	if (output_len < 0)
		fatal_error("%s: conversion error", filename);

	/* write file */
	fp = fopen(filename, "wb");
	if (fp == NULL)
		fatal_error("cannot write %s", filename);
	if (fwrite(output, output_len, 1, fp) != 1) {
		fclose(fp);
		remove(filename); /* "unlink" is less portable */
		fatal_error("%s: write error", filename);
	}
	fclose(fp);

	/* print summary */
	printf("%s: ", filename);
	int warnings = 0;

	/* issue a warning for samples - they may break on NTSC */
	/* TYPE S has FASTPLAY!=312, so it has been rejected earlier */
	if (ASAPInfo_GetTypeLetter(info) == 'D') {
		printf("WARNING: TYPE D");
		warnings++;
	}
	
	/* issue a warning if the 6502 code possibly reads the PAL/NTSC flag of GTIA
	   (LDA/LDX/LDY $D014, e.g. Ghostbusters.sap).
	   This is just a guess - false positives are possible and other code may be used for NTSC detection. */
	for (int i = 0; i < module_len - 2; i++) {
		if (module[i] >= 0xac && module[i] <= 0xae && module[i + 1] == 0x14 && module[i + 2] == 0xd0) {
			if (warnings++ > 0)
				printf(", ");
			printf("WARNING: possible PAL/NTSC detection code");
			break;
		}
	}

	ASAPInfo_Delete(info);
	if (warnings == 0)
		printf("ok");
	printf("\n");
}

int main(int argc, char **argv)
{
	bool usage = true;
	for (int i = 1; i < argc; i++) {
		const char *arg = argv[i];
		if (strcmp(arg, "--help") == 0) {
			usage = true;
			break;
		}
		if (strcmp(arg, "--version") == 0) {
			printf("sap2ntsc " ASAPInfo_VERSION "\n");
			return 0;
		}
		process_file(arg);
		usage = false;
	}
	if (usage) {
		printf(
			"Usage: sap2ntsc FILE.sap...\n"
			"Replaces FILE.sap with an NTSC version\n"
		);
	}
	return 0;
}
